### ISSUE TITLE
Parsers for API responses?

### ISSUE THREAD

**Issue by jkeljo** [sdwilsh/siobrultech-protocols#14]
Are you going to want to use the GEM API for your project?

I've been playing around with the API for auto-detecting the setup of the GEM (# of channels, etc.). The response formats are... interesting (comma-delimited hex values). I can do that parsing in my repo, or I can send you a PR here. What do you think? If the latter, any preferences as to how to set it up? I was thinking something along the lines of a class that takes in a `StreamReader` and `StreamWriter` and provides a bunch of `async` methods that make different requests and cajole the results into appropriate types.

**Comment by sdwilsh** [comment]
I'd like to, but it's probably not going to be a part of the MVP.

My case is more complicated because I'm using a serial bus.  I basically need a protocol that can handle packets and can handle the commands (same over HTTP or serial, as best I can tell).  That doesn't belong here, but what would be nice here is some kind of protocol that can handle the responses.  An `asyncio.StreamReader` and `asyncio.StreamWriter` could probably work too, I'd just have to figure out how to make it work with a serial connection somehow that also gets packet data.

**Comment by jkeljo** [comment]
I just realized (I commented on #13 as well) maybe I can treat the TCP/IP connection from the GEM as a serial bus and just send the commands down it rather than finding its IP and calling it back. It certainly seems to work that way on its end: if I open the setup page in my browser while that connection is active, it sends HTML down that connection as well as to my browser.

**Comment by jkeljo** [comment]
> what would be nice here is some kind of protocol that can handle the responses. An asyncio.StreamReader and asyncio.StreamWriter could probably work too, I'd just have to figure out how to make it work with a serial connection somehow that also gets packet data.

Yeah, the fun part is that the responses don't have enough information in them to determine what they're in response to. It's like you'd need a state machine or a queue of parsers to keep track of the responses you're expecting.

**Comment by sdwilsh** [comment]
Yup!  On the plus side, they have a command that asks to delay sending packets for N seconds, so you can do that, it'll finish the current packet, and then you have a window to send commands.

**Comment by jkeljo** [comment]
Ah, yeah, I see that one. OK, I'll take a shot at enabling this to send commands over the stream and hit you with a PR for discussion in a bit.
